model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +30,168 @@ pub struct CommandData {
 
 /// Data received when a user fills in a command option.
 ///
-/// Note: user, channel, role, and mentionable option types will be returned as
-/// a [`String`] option here.
-///
 /// Refer to [the discord docs] for more information.
 ///
 /// [the discord docs]: https://discord.com/developers/docs/interactions/slash-commands#interaction-applicationcommandinteractiondataoption
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-#[serde(untagged)]
-pub enum CommandDataOption {
-    String {
-        name: String,
-        value: String,
-    },
-    Integer {
-        name: String,
-        value: i64,
-    },
-    Boolean {
-        name: String,
-        value: bool,
-    },
-    SubCommand {
-        name: String,
-        #[serde(default)]
-        options: Vec<CommandDataOption>,
-    },
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CommandDataOption {
+    pub name: String,
+    pub value: CommandOptionValue,
```

Comment:
@zeylahellyer Is this struct def fine or should we keep the enum with serde untagged attribute?

Reply:
I'm not sure I understand exactly what you mean. Do you mean removing `CommandOptionValue` and adding new fields in the `enum CommandDataOption`, like this?
```rs
User {
    name: String,
    value: UserId,
},
// and same thing for Channel, Role, Mentionable, etc...
```

IMO this sounds suboptimal:
- the `name` field is duplicated in all variants, you need a `name()` function to access it, with all the downsides of this (ex can't pattern match on it)
- you can't use `#[serde(untagged)]` anyway, because you need to implement (de)serialization manually to handle the `type` field sent by Discord

The only upside I can think of is that the change becomes _less_ breaking. Forgive me if I'm missing something obvious.

Reply:
Yes, that is exactly what I meant. I am looking at it from the point of making it less breaking. I only mentioned the untagged variant to give a reference to the previous enum (I now realize how that made it sound confusing).

In my experience, I use both fields (name & value) whenever I access `CommandDataOption`. Thus, I just match over the entire enum. But that's just my personal preference.
e.g.
```rust
match option {
   CommandDataOption::User { name, value } => {/* do something*/},
   CommandDataOption::String { name, value } => {/* do something */},
   _ => {}
}
```

Most of the models in `twilight_model` are defined using enums (wherever there is a type field) so I would think this style should be followed in all future updates to be consistent.

Other than that, I do have a weird concern that has a very low probability of ever happening. Should Discord ever add a Command Option that doesn't have the name as a field, that would cause a problem.